drivers/mfd/rp1.c

```diff
@@ -63,6 +63,20 @@ static int rp1_probe(struct udevice *dev)
 	/* Turn on bus-mastering */
 	dm_pci_clrset_config16(dev, PCI_COMMAND, 0, PCI_COMMAND_MASTER | PCI_COMMAND_MEMORY);
 
+	/*
+	 * HACK: Set bar address in the correct order. RP1 driver in Linux Kernel rely on
+	 * the pci BAR configuration which initializes BARs based on BAR size.
+	 * This results the Linux set BARs in the same order on each boot.
+	 * U-boot does initialization without any sorting so the configuration
+	 * may be different from the Linux kernel.
+	 * Here BAR address configuration is set to match the Linux Kernel order
+	 * to avoid fails on address translation.
+	 * This should be changed after upstreaming RP1 driver to the Linux kernel
+	 * mainline.
+	 */
+	dm_pci_write_config32(dev, PCI_BASE_ADDRESS_1, 0);
+	dm_pci_write_config32(dev, PCI_BASE_ADDRESS_0, 0x40000);
+
```

Comment:
So my understanding is that it is required for Linux to probe PCIe and RP1, right? Wouldn't reset in Linux or here in u-boot help with that?

Reply:
Because Linux expecting PCIE to be configured in some state after firmware initialization. And doesn't reconfigure bars. It was lucky coincidence that Linux does initialization the same way firmware sets the values. but u-boot does init in the different pattern. that's why we see this HACK. Why it is so - well I can't tell you. I have no documentation for PCI 2712

Reply:
Yeah, design of RP1 driver on Linux is ... subpar.  It is working because of sheer luck (or because of assumption that prior bootloader will configure device in a certain way).